lang/wasm_instance_cache.hh

```diff
@@ -0,0 +1,326 @@
+    public:
+        cache_key_type(seastar::scheduling_group scheduling_group, const cql3::functions::function_name& name)
+            : scheduling_group(std::move(scheduling_group))
+            , name(name) {
+        }
+        bool operator==(const cache_key_type& other) const {
+            return scheduling_group == other.scheduling_group && name == other.name;
+        }
+    };
+    // temporary key with a reference to the function name to find elements in the cache
+    struct lookup_cache_key_type {
+        seastar::scheduling_group scheduling_group;
+        const cql3::functions::function_name& name;
+    public:
+        lookup_cache_key_type(seastar::scheduling_group scheduling_group, const cql3::functions::function_name& name) noexcept
+            : scheduling_group(std::move(scheduling_group))
+            , name(name) {
+        }
+        explicit lookup_cache_key_type(const cache_key_type& key) noexcept
+            : scheduling_group(key.scheduling_group)
+            , name(key.name) {
+        }
+        bool operator==(const cache_key_type& other) const {
+            return scheduling_group == other.scheduling_group && name == other.name;
+        }
+    };
+
+    struct cache_key_hash {
+        using is_transparent = void;
+
+        size_t operator()(const cache_key_type& v) const noexcept {
+            return utils::tuple_hash()(v.scheduling_group, v.name);
+        }
+        size_t operator()(const lookup_cache_key_type& v) const noexcept {
+            return utils::tuple_hash()(v.scheduling_group, v.name);
+        }
+    };
+
+    struct lru_entry_type {
+        lookup_cache_key_type key;
+        seastar::lowres_clock::time_point timestamp;
+        size_t instance_size;
+        lru_entry_type(const cache_key_type& key, seastar::lowres_clock::time_point timestamp, size_t size) noexcept :
+            key(key), timestamp(timestamp), instance_size(size) {}
+    };
+
+    struct cache_entry_type {
+        seastar::shared_mutex mutex;
+        seastar::lw_shared_ptr<wasm_instance> instance;
+        // iterator is invalid when the entry is being used (at that point, it is not in lru)
+        std::list<lru_entry_type>::iterator it;
+    };
+
+public:
+    using key_type = cql3::functions::function_name;
+    using value_type = lw_shared_ptr<wasm_instance>;
+
+private:
+    std::unordered_map<cache_key_type, cache_entry_type, cache_key_hash, std::equal_to<>> _cache;
+    std::list<lru_entry_type> _lru;
+    seastar::timer<seastar::lowres_clock> _timer;
+    // The instance in cache time out after up to 2*_timer_period.
+    seastar::lowres_clock::duration _timer_period;
+    size_t _total_size = 0;
+    size_t _max_size;
+
+public:
+    explicit instance_cache(size_t size, seastar::lowres_clock::duration timer_period)
+        : _timer([this] { return on_timer(); })
+        , _timer_period(timer_period)
+        , _max_size(size)
+    {
+        setup_metrics();
+        _timer.arm(_timer_period);
```

Comment:
Another way to go is to arm the timer for the last lru entry, instead of periodic calls. We don't need to go into that now.

Reply:
Yes, that gives us exact expiration, but on the other hand we would have much more timer wake-ups, so I wonder if that's not too expensive